x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -0,0 +1,142 @@
+    }
+
+    /**
+     * Asynchronously creates a new API key based off of the request and authentication
+     * @param authentication the authentication that this api key should be based off of
+     * @param request the request to create the api key included any permission restrictions
+     * @param listener the listener that will be used to notify of completion
+     */
+    public void createApiKey(Authentication authentication, CreateApiKeyRequest request, ActionListener<CreateApiKeyResponse> listener) {
+        ensureEnabled();
+        if (authentication == null) {
+            listener.onFailure(new IllegalArgumentException("authentication must be provided"));
+        } else {
+            final Instant created = clock.instant();
+            final Instant expiration = getApiKeyExpiration(created, request);
+            final SecureString apiKey = UUIDs.randomBase64UUIDSecureString();
+            final Version version = clusterService.state().nodes().getMinNodeVersion();
+            if (version.before(Version.V_7_0_0_alpha1)) { // TODO(jaymode) change to V6_5_0 on backport!
+                LOGGER.warn("nodes prior to the minimum supported version for api keys {} exist in the cluster; these nodes will not be " +
+                    "able to use api keys", Version.V_7_0_0_alpha1);
+            }
+
+            final char[] keyHash = Hasher.PBKDF2.hash(apiKey);
+            try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                builder.startObject()
+                    .field("doc_type", "api_key")
+                    .field("creation_time", created.toEpochMilli())
+                    .field("expiration_time", expiration == null ? null : expiration.toEpochMilli());
+
+                byte[] utf8Bytes = null;
+                try {
+                    utf8Bytes = CharArrays.toUtf8Bytes(keyHash);
+                    builder.field("api_key_hash").utf8Value(utf8Bytes, 0, utf8Bytes.length);
+                } finally {
+                    if (utf8Bytes != null) {
+                        Arrays.fill(utf8Bytes, (byte) 0);
+                    }
+                }
+
+                builder.array("role_descriptors", request.getRoleDescriptors())
+                    .field("name", request.getName())
+                    .field("version", version.id)
+                    .startObject("creator")
+                    .field("principal", authentication.getUser().principal())
+                    .field("metadata", authentication.getUser().metadata())
+                    .field("realm", authentication.getLookedUpBy() == null ?
+                        authentication.getAuthenticatedBy().getName() : authentication.getLookedUpBy().getName())
+                    .endObject()
+                    .endObject();
+                final IndexRequest indexRequest =
+                    client.prepareIndex(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, getApiKeyDocumentId(apiKey))
```

Comment:
We shouldn't use the api key value for the document id. If we still need it to be dependent on the api key ( so that we can get the document instead of search for authentication and invalidation ) we can use the hash, otherwise a random id would suffice

Reply:
> random id

The use of the key in the id was an oversight and I planned on using an autogenerated id for the document; these come from the `TimeBasedUUIDGenerator`.  Do you think this is ok or should we use a truly random value?

Reply:
Randomness doesn't buy us anything here so the requirements are the same as for any other auto-generated document id. A TimeBasedUUIDGenerator based one would be fine !